### PR TITLE
Make `TransportChannel#sendResponse` refcount-neutral

### PR DESCRIPTION
**Pull request by DaveCTurner**
Refcount-neutral methods are much easier on the reader because they mean
we can make obvious pairings between `incRef()` and `decRef()` calls.
This commit removes the implicit `decRef()` invocation from
`TransportChannel#sendResponse` to make it refcount-neutral.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
WIP because (a) CI might find more places to fix and (b) it can wait for in-flight work such as #103763 which I expect to have an impact too.

**Review by henningandersen — COMMENTED** [review]
This looks good to me. I do wonder how you came to the few specific production changes? I tried searching for usages of `sendResponse`, but going through them one by one seemed tedious.

File: test/framework/src/main/java/org/elasticsearch/transport/DisruptableMockTransport.java [hunk 1]
```diff
@@ -263,24 +264,28 @@ public String getProfileName() {
 
             @Override
             public void sendResponse(final TransportResponse response) {
+                response.mustIncRef();
+                final var releasable = Releasables.assertOnce(response::decRef);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/transport/DisruptableMockTransport.java:268]: `ReleasableRef` only really makes sense to save a few LoC when using try-with-resources blocks - if the lifecycle goes elsewhere, it's not really any neater than just passing the bare resource around as we do here.

File: test/framework/src/main/java/org/elasticsearch/core/ReleasableRef.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.core;
+
+/**
+ * Adapter to use a {@link RefCounted} in a try-with-resources block.
+ */
+public record ReleasableRef<T extends RefCounted>(T get) implements Releasable {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/core/ReleasableRef.java]: Fair point, yes, adjusted in a0b73fe272d.

File: test/framework/src/main/java/org/elasticsearch/core/ReleasableRef.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    @Override
+    public void close() {
+        get().decRef();
+    }
+
+    public static <T extends RefCounted> ReleasableRef<T> of(T value) {
+        return new ReleasableRef<>(value);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/core/ReleasableRef.java:34]: Yes with a0b73fe272d it's now a class with some more paranoia added. Shouldn't really matter if it's only used in try-with-resources blocks tho. Also everything that's refcounted protects against doing too many `decRef()` calls.

**Comment by DaveCTurner** [comment]
> I do wonder how you came to the few specific production changes?

Yeah I went through them all one-by-one. There's not really that many, only 25 in production code I think, and a decent fraction of those are sending `TransportResponse.Empty.INSTANCE` or some other concrete response type that has no meaningful refcounting.

**Comment by DaveCTurner** [comment]
Pro-tip: IntelliJ's "find" tab has a 'Remove' feature (⌫ on my keyboard) that lets you disregard matches one-by-one, or a whole tree of matches at once, which makes it easy to keep track of progress.

image

**Review by original-brownbear — APPROVED** [review]
LGTM makes sense to have this consistent, thanks for cleaning it up! :)

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java [hunk 4]
```diff
@@ -454,6 +454,9 @@ public void onResponse(SearchResponse response) {
                         channel.sendResponse(response);
                     } catch (Exception e) {
                         onFailure(e);
+                    } finally {
+                        // TODO - avoid the implicit incref elsewhere and then replace this whole thing with a ChannelActionListener
```
Comment (henningandersen) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java:458]: Is the counterpart not the explicit incref here:

https://github.com/elastic/elasticsearch/blob/1d7d9dc1eed2aeda08dd7a62a8a1893c5113a95b/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java#L162

I wonder if removing that (and fixing the todo) would work, but I imagine you having tried it out so curious about the result/where the implicit incref is?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java (hunk 4 above)
Comment (DaveCTurner) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java:458]: I think this one is also germane:

https://github.com/elastic/elasticsearch/blob/1d7d9dc1eed2aeda08dd7a62a8a1893c5113a95b/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupResponseTranslator.java#L245

But this code is also kinda branchy and I didn't want to spend too much time disentangling how the ownership of these refs is handed around. I opened https://github.com/elastic/elasticsearch/issues/104503 to track this debt.

Comment (DaveCTurner) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java:458]: NB I also don't trust the tests in this area, I tried making some obviously incorrect changes to this refcounting and everything still passed.

Comment (henningandersen) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java:458]: Sounds good, I also tried fiddling with the incref and it did not fail (which it sounds like it should have). But happy to leave it to the search team to fix that.

**Pull request merged**